DOCSP-12157: Fundamentals Extended JSON

ccho-mongodb opened the pull request:
## Pull Request Info

### Issue JIRA link:
https://jira.mongodb.org/browse/DOCSP-12157

### Docs staging link (requires sign-in on MongoDB Corp SSO):
https://docs-mongodbcom-staging.corp.mongodb.com/47fea5d/java/docsworker-xlarge/DOCSP-12157-fundamentals-json/fundamentals/data-formats/document-data-format-extended-json

### Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Does it render on staging correctly?
- [x] Are all the links working?
- [x] Are the staging links in the PR description updated?

cbush commented on source/fundamentals/data-formats/document-data-format-extended-json.txt:
```diff
@@ -0,0 +1,250 @@
+Overview
+--------
+
+JSON is a data format that represents data as objects that contain one or
```
This doesn't seem 100% technically accurate, since JSON doesn't have to encode an object and objects can be empty. Maybe to sharpen focus on the point you're making:

>JSON is a data format that encodes the values of objects, arrays, numbers, and so on. Extended JSON adds special syntax to encode type information for those values...

cbush reviewed (approved):
Looks good with one comment.

rozza commented on source/fundamentals/data-formats/document-data-format-extended-json.txt:
```diff
@@ -0,0 +1,250 @@
+.. code-block:: none
+   :copyable: false
+
+   {"_id": ObjectId("507f1f77bcf86cd799439011"), "myNumber": NumberLong(4794261)}
```
It might be good to explain how to add custom conversion of bson types.  

One such example can be found here: https://github.com/mongodb/mongo-kafka/blob/r1.3.0/src/main/java/com/mongodb/kafka/connect/source/json/formatter/SimplifiedJson.java

That takes the Relaxed output mode and also changes ObjectId's to output just hex strings, dates to be outputted as ISO date strings etc..

ccho-mongodb replied:
Sounds good, I'll add a section. Is this a frequently-asked question from the community/customers?

ccho-mongodb replied:
Addressed in f154afda39c6d5d83d850a77618fb0a8673b4eb5

rozza reviewed (changes requested):
Looks good a couple of suggestions

jyemin commented on source/fundamentals/data-formats/document-data-format-extended-json.txt:
```diff
@@ -0,0 +1,250 @@
+--------
+
+JSON is a data format that encodes the values of objects, arrays, numbers, and 
+so on. The **Extended JSON** format adds special syntax to JSON objects to 
```
I wouldn't call is special syntax.  Perhaps:

   The extended json format defines a reserved set of $-prefixed keys to represent....

ccho-mongodb replied:
Addressed in f154afda39c6d5d83d850a77618fb0a8673b4eb5

jyemin commented on source/fundamentals/data-formats/document-data-format-extended-json.txt:
```diff
@@ -0,0 +1,250 @@
+---------------------
+
+MongoDB Extended JSON represents BSON in different formats to handle
+specific use cases. These formats differ in the level of prioritization in
```
Can you try to re-word the sentence starting with "These formats".  I think I know what you mean but I think casual readers will not understand what you mean

ccho-mongodb replied:
Rewrote the paragraph f154afda39c6d5d83d850a77618fb0a8673b4eb5

jyemin commented on source/fundamentals/data-formats/document-data-format-extended-json.txt:
```diff
@@ -0,0 +1,250 @@
+       | For more information on this format, see the server manual page on :manual:`Data Types in the mongo shell </core/shell-types/>`.
+
+   * - **Strict**
+     - | Deprecated. This representation is the legacy format that fully conforms to the `JSON RFC <http://www.json.org/>`__ which allows any JSON parser to read the type information.
```
I'd prefer that we just don't document deprecated options.

ccho-mongodb replied:
I thought this was still the default format in which case I think we should include it for clarity. As an alternative, the outputMode could be "required" but that wouldn't be accurate. What would you suggest?

jyemin replied:
As of 4.0 RELAXED is the default.

ccho-mongodb replied:
Good to know, thanks. I'll remove it from the documentation.

ccho-mongodb replied:
Addressed in f154afda39c6d5d83d850a77618fb0a8673b4eb5

jyemin commented on source/fundamentals/data-formats/document-data-format-extended-json.txt:
```diff
@@ -0,0 +1,250 @@
+   507f1f77bcf86cd799439011 is type: org.bson.types.ObjectId
+   4794261 is type: java.lang.Long
+
+To construct an instance of a ``Document`` from Extended JSON, you can call
```
Document.parse() is going to be the way that almost everyone will parse JSON.  Should we lead with that and leave JsonReader usage to the end?

Same goes for Document.toJson() vs. JsonWriter

ccho-mongodb replied:
Ah, I thought it could be important to explain how the reader or writer works first to provide them with the context of how it works. I don't have a strong preference here; I'll update to present the Document methods first and see how it looks.

ccho-mongodb replied:
Addressed in f154afda39c6d5d83d850a77618fb0a8673b4eb5 and b4e42b9e7d6162577786543786ae09e64207102b

ccho-mongodb commented on source/fundamentals/data-formats/document-data-format-extended-json.txt:
```diff
@@ -0,0 +1,250 @@
+Overview
+--------
+
+JSON is a data format that encodes the values of objects, arrays, numbers, and 
```
Addressed in f154afda39c6d5d83d850a77618fb0a8673b4eb5

ccho-mongodb commented:
Ready for another round of reviews.
Changes included in f154afda39c6d5d83d850a77618fb0a8673b4eb5 and b4e42b9e7d6162577786543786ae09e64207102b :

- Updated headings
- Updated order of presentation of Reader/Writers
- Added section on Custom BSON Type Conversion
- Removed deprecated strict JSON format
- Made requested word choice changes

jyemin commented on source/fundamentals/data-formats/document-data-format-extended-json.txt:
```diff
@@ -111,26 +115,50 @@ corresponds to the format of the example you want to see:
+   String ejsonStr = "{ \"_id\": { \"$oid\": \"507f1f77bcf86cd799439011\"}," +
+                     "\"myNumber\": {\"$numberLong\": \"4794261\" }}}";
+
+   Document doc = new Document(Document.parse(ejsonStr));
```
Just 

     Document doc = Document.parse(ejsonStr);

is sufficient.

ccho-mongodb replied:
Addressed in https://github.com/mongodb/docs-java/pull/37/commits/47fea5dd750da5601c6ba188a3bf768aa5b9a954

jyemin commented on source/fundamentals/data-formats/document-data-format-extended-json.txt:
```diff
@@ -164,47 +192,48 @@ The output of this code example resembles the following:
-   Document doc = new Document(Document.parse(ejsonStr));
-   System.out.println(doc);
+You can write an Extended JSON string from an instance of ``Document`` or
+``BsonDocument`` by calling the ``toJson()`` method, passing it an instance of
```
optionally passing

ccho-mongodb replied:
Addressed in https://github.com/mongodb/docs-java/pull/37/commits/47fea5dd750da5601c6ba188a3bf768aa5b9a954

jyemin commented on source/fundamentals/data-formats/document-data-format-extended-json.txt:
```diff
@@ -164,47 +192,48 @@ The output of this code example resembles the following:
+``JsonWriterSettings`` to specify the Extended JSON format.
 
-The output of the code above resembles the following:
+In this example, we output the Extended JSON in the "Shell" format.
```
Let's not show Shell format.  How about relaxed instead?

ccho-mongodb replied:
Addressed in https://github.com/mongodb/docs-java/pull/37/commits/47fea5dd750da5601c6ba188a3bf768aa5b9a954

jyemin commented on source/fundamentals/data-formats/document-data-format-extended-json.txt:
```diff
@@ -164,47 +192,48 @@ The output of this code example resembles the following:
-class. To construct an instance of ``JsonWriter``, pass a
-subclass of a Java ``Writer`` to specify how you want to output the Extended
-JSON. You can optionally pass a :java-docs:`JsonWriterSettings </apidocs/bson/org/bson/json/JsonWriterSettings.html>`
+You can also read an Extended JSON string into Java objects using the BSON
```
This is the opposite of what it does.  It writes JSON, not reads it.

ccho-mongodb replied:
Addressed in https://github.com/mongodb/docs-java/pull/37/commits/47fea5dd750da5601c6ba188a3bf768aa5b9a954

rozza reviewed (approved):
LGTM! Thanks @ccho-mongodb
